File: codility/go/3-time_complexity.go
package golang

func frogJmp(x int, y int, d int) int {
	dist := y - x
	if dist%d == 0 {
		return dist / d
	} else {
		return dist/d + 1
	}
}

func permMissingElem(list []int) int {
	l := len(list)
	// use Gauss's formula to calculate the sum of the first n positive integers
	// then subtract the sum of the list to get the missing number
	return (l+1)*(l+2)/2 - sum(list)
}

func sum(list []int) int {
	total := 0
	for _, n := range list {
		total += n
	}
	return total
}

func permMissingElem2(list []int) int {
	l := len(list)
	// use XOR to find the missing number
	// XOR will cancel out pairs of numbers, when using index + 1 as the other number in a pair
	// the missing number will be left over
	// the array does not need to be sorted, but this only works if there is only one missing number
	num := 0
	for i := 0; i < l; i++ {
		num ^= list[i] ^ (i + 1)
	}
	return num ^ (l + 1)
}

func tapeEquilibrium(list []int) int {
	left := list[0]
	right := list[1]
	for i := 2; i < len(list); i++ {
		right += list[i]
	}
	best := right - left
	if best < 0 {
		best = -best
	}
	for i := 1; i < len(list)-1; i++ {
		left += list[i]
		right -= list[i]
		diff := right - left
		if diff < 0 {
			diff = -diff
		}
		if diff < best {
			best = diff
		}
	}

	return best
}

File: codility/py/3-time_complexity.py
import math


def frog_jump(pos, target_pos, jump_distance):
    return math.ceil((target_pos - pos) / jump_distance)


assert frog_jump(10, 85, 30) == 3
assert frog_jump(10, 100, 30) == 3
assert frog_jump(10, 100, 10) == 9


def missing_element(A):
    # XOR all elements with their expected positions
    res = 0
    for i in range(len(A)):
        res ^= i + 1
        res ^= A[i]

    # There is only one number missing (per task spec)
    # XOR result with replacement number (len(A)+1)
    return res ^ (len(A) + 1)


assert missing_element([2, 3, 1, 5]) == 4
assert missing_element([6, 1, 4, 2, 3]) == 5
assert missing_element([5, 2, 4, 3]) == 1
assert missing_element([4, 1, 2, 3]) == 5


def tape_equilibrium(A):
    left, right = 0, sum(A)
    # best = right doesn't work for negative numbers
    best = float('inf')

    for p in range(1, len(A)):
        n = A[p - 1]
        left += n
        right -= n
        best = min(best, abs(left - right))

    return best


assert tape_equilibrium([3, 1, 2, 4, 3]) == 1
assert tape_equilibrium([1, 1]) == 0
assert tape_equilibrium([-1000, 1000]) == 0
assert tape_equilibrium([1000, -1000]) == 2000
assert tape_equilibrium([-10, -20, -30, -40, 100]) == 0
